modules/core/src/has_non_zero.simd.hpp

```diff
@@ -0,0 +1,325 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+#include "precomp.hpp"
+
+namespace cv {
+
+typedef bool (*HasNonZeroFunc)(const uchar*, size_t);
+
+
+CV_CPU_OPTIMIZATION_NAMESPACE_BEGIN
+
+HasNonZeroFunc getHasNonZeroTab(int depth);
+
+
+#ifndef CV_CPU_OPTIMIZATION_DECLARATIONS_ONLY
+
+template<typename T>
+inline bool hasNonZero_(const T* src, size_t len )
+{
+    bool res = false;
+    if (len > 0)
+    {
+        size_t i=0;
+        #if CV_ENABLE_UNROLLED
+        for(; !res && (i+4 <= len); i += 4 )
+            res |= ((src[i] | src[i+1] | src[i+2] | src[i+3]) != 0);
+        #endif
+        for( ; !res && (i < len); i++ )
+            res |= (src[i] != 0);
+    }
+    return res;
+}
+
+template<>
+inline bool hasNonZero_(const float* src, size_t len )
+{
+    bool res = false;
+    if (len > 0)
+    {
+        size_t i=0;
+        if (sizeof(float) == sizeof(unsigned int))
+        {
+            #if CV_ENABLE_UNROLLED
+            typedef unsigned int float_as_uint_t;
+            const float_as_uint_t* src_as_ui = reinterpret_cast<const float_as_uint_t*>(src);
+            for(; !res && (i+4 <= len); i += 4 )
+            {
+                const float_as_uint_t gathered = (src_as_ui[i] | src_as_ui[i+1] | src_as_ui[i+2] | src_as_ui[i+3]);
+                res |= ((gathered<<1) != 0);//remove what would be the sign bit
+            }
+            #endif
+        }
+        for( ; !res && (i < len); i++ )
+            res |= (src[i] != 0);
+    }
+    return res;
+}
+
+template<>
+inline bool hasNonZero_(const double* src, size_t len )
+{
+    bool res = false;
+    if (len > 0)
+    {
+        size_t i=0;
+        if (sizeof(double) == sizeof(uint64_t))
+        {
+            #if CV_ENABLE_UNROLLED
+            typedef uint64_t double_as_uint_t;
+            const double_as_uint_t* src_as_ui = reinterpret_cast<const double_as_uint_t*>(src);
+            for(; !res && (i+4 <= len); i += 4 )
+            {
+                const double_as_uint_t gathered = (src_as_ui[i] | src_as_ui[i+1] | src_as_ui[i+2] | src_as_ui[i+3]);
+                res |= ((gathered<<1) != 0);//remove what would be the sign bit
+            }
+            #endif
+        }
+        for( ; !res && (i < len); i++ )
+            res |= (src[i] != 0);
```

Comment:
<<1 is missing here

Reply:
I don't think so. Native comparison between a float (or double) and zero already handles the sign bit.